File: 8-one more time/ruby/interpol.rb
require 'matrix'

def polyfit(x, y, degree)
  x = x.to_a
  y = y.to_a

  m = Matrix.build(x.size, [degree, x.size].min) do |i, j|
    x[i]**j
  end

  b = (m.transpose * m).inverse * m.transpose * Vector.elements(y)

  return b.to_a
end

# Example usage:
x = Vector[0, 1, 2]
y = Vector[2, 2, 2]
degree = 2

coefficients = polyfit(x, y, degree)
puts "Polynomial coefficients: #{coefficients}"

File: 8-one more time/python 2/interpol.py
import numpy as np;

def polyfit(x, y, degree):
    """
    Fits a polynomial of given degree to the data points (x, y) using matrix calculations.
    
    Arguments:
    x -- array-like object representing the x-coordinates of the data points
    y -- array-like object representing the y-coordinates of the data points
    degree -- degree of the polynomial to fit
    
    Returns:
    coefficients -- array of polynomial coefficients in decreasing order of degree
    """
    x = np.asarray(x)
    y = np.asarray(y)
    
    m = np.ones((len(x), min(degree, len(x))))
    for i in range(min(degree, len(x))):
        m[:, i] *= x**i

    b = np.matmul(np.linalg.pinv(m), y)
    
    return b

# Example usage:
x = np.array([0, 1, 2])
y = np.array([3, 6, 13])
degree = 3

coefficients = polyfit(x, y, degree)
print coefficients